runners/apex/src/main/java/org/apache/beam/runners/apex/translation/utils/ApexTimerInternals.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.apex.translation.utils;
+
+import com.datatorrent.netlet.util.Slice;
+import com.esotericsoftware.kryo.DefaultSerializer;
+import com.esotericsoftware.kryo.serializers.JavaSerializer;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import org.apache.beam.runners.core.StateNamespace;
+import org.apache.beam.runners.core.TimerInternals;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.joda.time.Instant;
+
+/**
+ * An implementation of Beam's {@link TimerInternals}.
+ *
+ * <p>Assumes that the current key is set prior to accessing the state.<br>
+ * This implementation stores timer data in heap memory and is serialized
+ * during checkpointing, it will only work with a small number of timers.
+ * @param <K>
+ */
+@DefaultSerializer(JavaSerializer.class)
+public class ApexTimerInternals<K> implements TimerInternals, Serializable {
+
+  private Map<Slice, Set<Slice>> activeTimers = new HashMap<>();
```

Comment:
You may want a `NavigableSet` in case many windows become active on a key (I think this would require the watermark to stall for a bit).

Given the way you implement this - with the impetus for processing coming separately and just saving the timer data here - could you easily support cancel in the same manner as the direct runner? https://github.com/apache/beam/blob/master/runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java#L223